### PR TITLE
Student multiple choice data hookup

### PR DESCRIPTION
**Pull request by nkiruka**
## Summary
Hooks up first react table (already built) to our redux store (already built) using two selector functions (new) and a container component (new). 

Table that got hooked up: individual student's responses for the multiple choice questions in a single assessment.

screen shot 2018-06-06 at 10 36 24 am

## User impact
None without internal flag `assessmentTab`.

## Implementation details
Implemented two selector functions - one for the structure of the assessment, and one for the student responses. Both of these have redux tests.
Selector functions are used in a new container component that displays a table for each student we have in the redux store.

We had to modify our react tables a bit to get this working (mostly delete code) and be more clear about our prop types, although there is probably still more to do here. Nkiru has a PR ready for after this that includes renaming of some components and comments for each of our components since there are so many types of tables. We've also updated the storybook data for any of the components we've changed.

For re-naming, we want to consistently use the term "responses" to mean student responses to an assessment, and "structure" to mean the questions/answers/structure of the assessment that every student sees. We've done this for the selectors and some props, but we're not all the way there yet.
The [Glossary](https://docs.google.com/document/d/1hJFCogMy5fgK1WOxcmWVqtGgOnEdDeKZnC6SkRFHQ-I/edit?usp=sharing) document includes re-named tables and prop types.

We've left ourselves some todos in this PR that we indent to follow up with in other PRs.

@caleybrock and @nkiruka paired on this PR.

### CODE REVIEW

File: apps/src/templates/sectionAssessments/MultipleChoiceByStudentSection.jsx [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  static propTypes = {
+    multipleChoiceStructure: PropTypes.arrayOf(questionStructurePropType),
+    studentAnswerData: PropTypes.arrayOf(studentAnswerDataPropType),
+  };
+
+  render() {
+    const {multipleChoiceStructure, studentAnswerData} = this.props;
+    return (
+      <div>
+        <h2>Multiple choice answers by student section</h2>
```
Comment (islemaster) [inline apps/src/templates/sectionAssessments/MultipleChoiceByStudentSection.jsx:20]: Localization now, or later?

**Comment by caleybrock** [comment]
Adding Maddie because I just realized Brad probably has a backlog of reviews :)

File: apps/src/templates/sectionAssessments/MultipleChoiceByStudentSection.jsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import {
+  getMultipleChoiceStructureForCurrentAssessment,
+  getStudentMCResponsesForCurrentAssessment,
+} from './sectionAssessmentsRedux';
+import { connect } from 'react-redux';
+
+class MultipleChoiceByStudentSection extends Component {
+  static propTypes = {
+    multipleChoiceStructure: PropTypes.arrayOf(questionStructurePropType),
+    studentAnswerData: PropTypes.arrayOf(studentAnswerDataPropType),
```
Comment (islemaster) [inline apps/src/templates/sectionAssessments/MultipleChoiceByStudentSection.jsx:13]: Naming nit: Words like "structure" and "data" don't tell me a whole lot about the contents of these variables, and I'd omit them when possible.  I wonder if something like `questions` and `studentAnswers` is sufficient in this context?

File: apps/src/templates/sectionAssessments/MultipleChoiceByStudentSection.jsx [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import React, {Component, PropTypes} from 'react';
+import StudentAssessmentOverviewTable from './StudentAssessmentOverviewTable';
+import { studentAnswerDataPropType, questionStructurePropType } from './assessmentDataShapes';
+import {
+  getMultipleChoiceStructureForCurrentAssessment,
+  getStudentMCResponsesForCurrentAssessment,
+} from './sectionAssessmentsRedux';
+import { connect } from 'react-redux';
+
+class MultipleChoiceByStudentSection extends Component {
```
Comment (islemaster) [inline apps/src/templates/sectionAssessments/MultipleChoiceByStudentSection.jsx:10]: What is a "student section"?  Do you mean "by student by section" in this name?

File: apps/src/templates/sectionAssessments/assessmentDataShapes.js [hunk 4]
```diff
@@ -18,3 +12,29 @@ export const questionDataPropType = PropTypes.shape({
   answers: PropTypes.arrayOf(answerDataPropType),
   notAnswered: PropTypes.number.isRequired,
 });
+
+// Shapes for single student multiple choice tables
+
+// Represents an individual students response to a single
+// multiple choice question.
```
Comment (islemaster) [inline apps/src/templates/sectionAssessments/assessmentDataShapes.js:19]: Great comments on these PropTypes explaining what they're used for!

Nit: The missing apostrophe sort of changes the meaning. Should be `individual student's response`.  Could also be terser: "Represents one student's response to one multiple choice question."

File: apps/src/templates/sectionAssessments/assessmentDataShapes.js [hunk 5]
```diff
@@ -18,3 +12,29 @@ export const questionDataPropType = PropTypes.shape({
...
+export const studentResponsePropType = PropTypes.shape({
+  isCorrect: PropTypes.bool,
+  answers: PropTypes.string,
+});
+
+// Represents a single student and a set of there answers for
+// a single assessment's multiple choice questions
```
Comment (islemaster) [inline apps/src/templates/sectionAssessments/assessmentDataShapes.js:26]: Spelling: `there/their`.

Would it be accurate to say "Represents one student and _all_ of their multiple-choice answers for one assessment"?

With all of these prop types, let's establish some shared terminology.  Having a little glossary would help, something like this:

- **answer** - a single selected choice in a multiple-choice question.
- **response** - a set of selected choices in a multiple-choice question.
- ...

File: apps/src/templates/sectionAssessments/assessmentDataShapes.js [hunk 6]
```diff
@@ -18,3 +12,29 @@ export const questionDataPropType = PropTypes.shape({
...
+export const studentAnswerDataPropType = PropTypes.shape({
+  id:  PropTypes.string,
+  name: PropTypes.string,
+  studentAnswers: PropTypes.arrayOf(studentResponsePropType).isRequired,
+});
+
+// Represents a single multiple choice question structure
+export const questionStructurePropType = PropTypes.shape({
```
Comment (islemaster) [inline apps/src/templates/sectionAssessments/assessmentDataShapes.js:34]: What about `multipleChoiceQuestionPropType`?

File: apps/src/templates/sectionAssessments/assessmentDataShapes.js [hunk 7]
```diff
@@ -18,3 +12,29 @@ export const questionDataPropType = PropTypes.shape({
...
+// Represents an individual students response to a single
+// multiple choice question.
+export const studentResponsePropType = PropTypes.shape({
+  isCorrect: PropTypes.bool,
+  answers: PropTypes.string,
+});
+
+// Represents a single student and a set of there answers for
+// a single assessment's multiple choice questions
+export const studentAnswerDataPropType = PropTypes.shape({
```
Comment (islemaster) [inline apps/src/templates/sectionAssessments/assessmentDataShapes.js:27]: Maybe `studentWithResponsesPropType`?

File: apps/src/templates/sectionAssessments/MultipleChoiceByStudentSection.jsx (hunk 1 above)
Comment (caleybrock) [reply apps/src/templates/sectionAssessments/MultipleChoiceByStudentSection.jsx:20]: This heading isn't permanent, but a helpful thing during prototyping.

File: apps/src/templates/sectionAssessments/MultipleChoiceByStudentSection.jsx (hunk 3 above)
Comment (caleybrock) [reply apps/src/templates/sectionAssessments/MultipleChoiceByStudentSection.jsx:10]: MultipleChoiceByStudent, section (aka container). Maybe it's helpful to call it container and add a comment.

**Comment by caleybrock** [comment]
Great feedback on naming. We have a PR on the way, Nkiru will link to a glossary doc / plan for this once it exists. We're going to merge for now and do a follow up PR with a bunch of name changes to make everything more clear.

**Pull request merged**